generator/internal/descriptor_utils.cc

```diff
@@ -210,13 +211,79 @@ std::string DefaultIdempotencyFromHttpOperation(
       std::exit(1);
   }
 }
+
+std::string FormatClassCommentsFromServiceComments(
+    google::protobuf::ServiceDescriptor const& service) {
+  google::protobuf::SourceLocation service_source_location;
+  if (service.GetSourceLocation(&service_source_location) &&
+      !service_source_location.leading_comments.empty()) {
+    return absl::StrReplaceAll(
+        absl::StrCat(
+            "/**\n *",
+            absl::StrReplaceAll(service_source_location.leading_comments,
+                                {{"\n", "\n * "}}),
+            "\n", " */"),
+        {{"*  ", "* "}});
+  }
+  GCP_LOG(FATAL) << __FILE__ << ":" << __LINE__ << ": " << service.full_name()
+                 << " no leading_comments to format.\n";
+  return {};
+}
+
+enum class MethodParameterStyle { kApiMethodSignature, kProtobufReqeust };
+
+std::string FormatMethodCommentsFromRpcComments(
+    google::protobuf::MethodDescriptor const& method,
+    MethodParameterStyle parameter_style) {
+  google::protobuf::SourceLocation method_source_location;
+  if (method.GetSourceLocation(&method_source_location) &&
+      !method_source_location.leading_comments.empty()) {
+    std::vector<std::pair<std::string, std::string>> parameter_comments;
+    if (parameter_style == MethodParameterStyle::kApiMethodSignature) {
+      auto method_signature_extension =
+          method.options().GetRepeatedExtension(google::api::method_signature);
+      for (auto const& signature : method_signature_extension) {
+        google::protobuf::Descriptor const* input_type = method.input_type();
+        std::vector<std::string> parameters = absl::StrSplit(signature, ",");
+        for (auto const& parameter : parameters) {
+          google::protobuf::FieldDescriptor const* parameter_descriptor =
+              input_type->FindFieldByName(parameter);
+          google::protobuf::SourceLocation loc;
+          parameter_descriptor->GetSourceLocation(&loc);
+          parameter_comments.emplace_back(
+              parameter,
+              absl::StrReplaceAll(loc.leading_comments, {{"\n", "\n   * "}}));
+        }
+      }
+    } else {
+      google::protobuf::Descriptor const* input_type = method.input_type();
+      parameter_comments.emplace_back(
+          "request",
+          absl::StrCat("`", ProtoNameToCppName(input_type->full_name()), "`"));
+    }
+    std::string parameter_comment_string;
+    for (auto const& param : parameter_comments) {
+      parameter_comment_string +=
+          absl::StrFormat("   * @param %s %s\n", param.first, param.second);
+    }
+    return absl::StrCat(
```

Comment:
Here too, consider breaking this call up into a few lines to make it easier for readers to see what transformations are taking place and in what order.